{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"EvolutionParentalCare_Bendeskyetal_2017_partial.ipynb","provenance":[{"file_id":"1to5Jk8sp5yTr0SW2PkB9xkx7BDTMyZDp","timestamp":1613753290440}],"collapsed_sections":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.5"}},"cells":[{"cell_type":"markdown","metadata":{"id":"aNYpL-KmwFG8"},"source":["# **The genetic basis of parental care evolution in monogamous mice.** \n","### Andres Bendesky, Young-Mi Kwon,  Jean-Marc Lassance,  Caitlin L. Lewarch, Shenqin Yao, Brant K. Peterson, Meng Xiao He,  Catherine Dulac,  & Hopi E. Hoekstra\n","> *Nature* **2017**\n","\n","> **ABSTRACT:** Parental care is essential for the survival of mammals, yet the mechanisms underlying its evolution remain largely unknown. Here we show that two sister species of mice, Peromyscus polionotus and Peromyscus maniculatus, have large and heritable differences in parental behaviour. Using quantitative genetics, we identify 12 genomic regions that affect parental care, 8 of which have sex-specific effects, suggesting that parental care can evolve independently in males and females. Furthermore, some regions affect parental care broadly, whereas others affect specific behaviours, such as nest building. Of the genes linked to differences in nest-building behaviour, vasopressin is differentially expressed in the hypothalamus of the two species, with increased levels associated with less nest building. Using pharmacology in Peromyscus and chemogenetics in Mus, we show that vasopressin inhibits nest building but not other parental behaviours. Together, our results indicate that variation in an ancient neuropeptide contributes to interspecific differences in parental care."]},{"cell_type":"code","metadata":{"cellView":"form","id":"1MegNozwwQWm","colab":{"base_uri":"https://localhost:8080/","height":343},"executionInfo":{"status":"ok","timestamp":1613697648991,"user_tz":300,"elapsed":419,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}},"outputId":"110cc021-0ad6-484e-d40a-8a14a55643d3"},"source":["# @title Monogomous and Permiscuous voles\n","%%html\n","<div align='center'>\n","  <img src='https://drive.google.com/uc?export=view&id=15MZd0zThC4KbFyPUau8CoHZNtb6XHPTS' style=\"width: 85vw\"></div>\n","\n"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/html":["<div align='center'>\n","  <img src='https://drive.google.com/uc?export=view&id=15MZd0zThC4KbFyPUau8CoHZNtb6XHPTS' style=\"width: 85vw\"></div>\n"],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"markdown","metadata":{"id":"7JCDFgU1wFHD"},"source":["### In preparation for this classroom session you read the paper \"*The genetic basis of parental care evolution in monogamous mice*\" as well as the \"News and Views\" commentary \"*How to build a better dad*\" written by Steven Phelps.<br>\n","---\n","# **Today** you will work directly with data from this paper.\n"]},{"cell_type":"code","metadata":{"id":"4IkKVtHE2UNq","cellView":"form","executionInfo":{"status":"ok","timestamp":1613759347645,"user_tz":300,"elapsed":494,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}}},"source":["#@title __execute__ this cell to prepare your notebook environment with the necessary python packages \n","\n","# and mount your google drive\n","\n","import numpy as np\n","import urllib.request\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","import matplotlib\n","import seaborn as sns\n","import scipy\n","from IPython.display import display\n","\n","plt.rcParams.update({'font.size': 18})\n","\n","# from google.colab import drive\n","# drive.mount('/content/gdrive')\n","\n","def create_figure():\n","  hfig = plt.figure()\n","  ax = hfig.add_axes([0.1,0.1,0.8,0.8])\n","  return ax"],"execution_count":15,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"wpcSfWv2sdxK"},"source":["---\n","#Before working with the data, let's reflect on some of the important aspects of the experimental design.\n","\n","### Diversity of parental care behavior is widespread in the animal kingdom. What was special about these two vole species in particular that enabled the researchers to interrogate the genetic basis of parental care behaviors?"]},{"cell_type":"markdown","metadata":{"id":"Rf1Flgs1sdxK"},"source":["__Type your answer__ in this cell markdown cell"]},{"cell_type":"markdown","metadata":{"id":"X8IjyvF_okQx"},"source":["### Below are 4 videos that demonstrate the parental behavior of mothers and fathers of each vole species. Each video is long (~20min) so you won't have time to watch them now, but you can flip through a few different points in each video really quick to get a general sense of differences. You can revisit these videos later on your own if you like. \n","> ***note*** Do not spend more than 5 minutes looking at these videos here during class.\n"]},{"cell_type":"code","metadata":{"id":"tioigantjfAy","colab":{"base_uri":"https://localhost:8080/","height":357},"cellView":"form","executionInfo":{"status":"ok","timestamp":1613697649747,"user_tz":300,"elapsed":1166,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}},"outputId":"831c627d-7ef5-4435-c5d6-1dd2387daf49"},"source":["#@title Parental Behavior of a Peromyscus polionotus father:\n","%%html\n","<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM129_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/html":["<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM129_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"eGalP352j4w0","colab":{"base_uri":"https://localhost:8080/","height":357},"cellView":"form","executionInfo":{"status":"ok","timestamp":1613697649748,"user_tz":300,"elapsed":1159,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}},"outputId":"a871d624-5631-4df0-ad50-b171d2b101d6"},"source":["#@title Parental Behavior of a Peromyscus maniculatus father:\n","%%html\n","<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM130_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/html":["<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM130_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"Q_WhGqLWnHa8","colab":{"base_uri":"https://localhost:8080/","height":357},"cellView":"form","executionInfo":{"status":"ok","timestamp":1613697649748,"user_tz":300,"elapsed":1151,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}},"outputId":"5786012b-5af4-4974-fa44-60e996014645"},"source":["#@title Parental Behavior of a Peromyscus polionotus mother:\n","%%html\n","<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM131_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/html":["<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM131_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"_JMTv8Nkm7KV","colab":{"base_uri":"https://localhost:8080/","height":357},"cellView":"form","executionInfo":{"status":"ok","timestamp":1613697649749,"user_tz":300,"elapsed":1145,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}},"outputId":"1660aa62-6ffd-44eb-8201-3473c18cb068"},"source":["#@title Parental Behavior of a Peromyscus maniculatus mother:\n","%%html\n","<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM132_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/html":["<div align=\"center\">\n","<video width=600 controls>\n","      <source src=\"https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM132_ESM.mp4\" type=\"video/mp4\">\n","</video>\n","</div>"],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"markdown","metadata":{"id":"gPbh8ARVwFHG"},"source":["-----\n","# Let's work with the data!\n","* For each Figure, code has been written for you that imports the raw data that the authors provided with the online version of [the research article](https://www.nature.com/articles/nature22074#Sec28).\n","* The data will be loaded into a \"*pandas*\" [dataframe](https://realpython.com/pandas-dataframe/) format.\n"," *  Essentially, dataframes are like 'Microsoft Excell' sheets that we can access with python code and that have specific functions associated with them that makes data analysis easier. They have column headers (labels) and row indices to organise the data.\n","----\n","\n"]},{"cell_type":"markdown","metadata":{"id":"X7mXTdxTkUTq"},"source":["\n","# <font color=green> Figure 1</font> :\n"]},{"cell_type":"code","metadata":{"id":"SUA1Yi0MfGj2","cellView":"form"},"source":["#@title **execute** this cell to IMPORT & FORMAT the RAW DATA <br> (if you are interested seeing the code that makes that happen, you can double click this cell)\n","\n","# URL of data to download\n","data_url = 'https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM110_ESM.xlsx'\n","\n","# Get the data and save it locally as \"data.xls\"\n","data, headers = urllib.request.urlretrieve(data_url, './data.xls')\n","\n","xls = pd.ExcelFile(data)\n","df_NestQuality = pd.read_excel(xls, sheet_name = 'panel C', skiprows = 1)\n","df_Licking = pd.read_excel(xls, sheet_name = 'panel D', skiprows = 1)\n","df_Huddling = pd.read_excel(xls, sheet_name = 'panel E', skiprows = 1)\n","df_Retrieving = pd.read_excel(xls, sheet_name = 'panel F', skiprows = 1)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4yB1eKZysdxO"},"source":["### You now have a dataframe for the data underlying each of the Panels B-E of Figure 1 with the following names: \n","- df_Licking\n","- df_NestQuality\n","- df_Huddling\n","- df_Retrieving\n","\n","### Like each Figure panel, the data in the dataframe is grouped into 4 categories corresponding to the mothers and fathers of each species... <br>\n","### Pick a behavior and Take a look!\n"]},{"cell_type":"code","metadata":{"id":"QeKBZokqsdxO","executionInfo":{"status":"ok","timestamp":1613762434009,"user_tz":300,"elapsed":424,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}}},"source":["# Type the name of one of the four dataframes here and execute the cell to view\n","# (how is the appearance of the result different when you use the print() function to view?)\n"],"execution_count":28,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"TaMipPmUcX-Q"},"source":["#### You can also just display the first few lines of a dataframe by using the head( ) method : \n","`dataframe_name.head()` <br>\n","Try this in the cell below"]},{"cell_type":"code","metadata":{"id":"Ad0uqFyPct3A"},"source":["# display only the first few lines of the dataframe you viewed above\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"prujJzzqmqix"},"source":["### Did you notice 'NaN' at the end of some of the columns? <br>\n","### What does this mean and why are there NaN in the dataframe? <br>\n"," \n","> __hint__ look at the 'n' noted at the bottom of the Figure 1. "]},{"cell_type":"markdown","metadata":{"id":"uqbIuu1Emp8S"},"source":["__Type your answer__ in this cell markdown cell"]},{"cell_type":"markdown","metadata":{"id":"p-j1ED19KfYy"},"source":["### Let's plot the data to visualize it.\n","* [Seaborn](https://seaborn.pydata.org/) is a statistical data visualization package that has a lot of great plotting functions for dataframes.\n","* run the cell below to import the seaborn package as \"`sns`\" "]},{"cell_type":"code","metadata":{"id":"82xrhd17Kh1C"},"source":["import seaborn as sns"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"8vVOdhEqK306"},"source":["* Since the authors used boxplots to display the data, let's do the same for now. <br> Seaborn has a [boxplot( )](https://seaborn.pydata.org/generated/seaborn.boxplot.html) function. <br> To use it, the most basic format would be `sns.boxplot(data = dataframe_name)`\n","\n"," * **hint** : put a '`;`' at the end of the line of code to suppress text output associated with the plot\n","\n"," * This function automatically finds the names of the columns in the dataframe to use as axis labels."]},{"cell_type":"code","metadata":{"id":"qDt5P7bhtVGy"},"source":["# Make a boxplot of the dataframe for the behavior you chose to work with\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"P2ssfYQ2sdxi"},"source":["### Great! Now that you understand better how your dataframe is organized and what data it contains,  let's visualize a comparison that was not explicitly made in the paper : \n","### **The difference in parental care behavior betwee sexes (ignoring species identity).**\n","\n","### First we need to collect the data in a new way : We need a list of the data for all females (ie. \"mothers\" of both species combined) and a list of the data for all males (\"fathers\").\n","\n","* To get the data from a whole column of a dataframe, \"*index*\" it using the column name (which is a string).\n","* Remember that indices are placed inside brackets ( `[ ]` ).\n","* The contents of a column can be assigned to a new variable name.\n","\n","> ***note*** for the following sections, decide on one behavior to look at (licking, nest quality, etc) and apply the instructions to the corresponding dataframe of your choice. (in other words, be consistent about which dataframe you are indexing) \n"]},{"cell_type":"code","metadata":{"id":"J0pKqiOvFvdO"},"source":["# Assign the data for P.maniculatus fathers from your dataframe of choice to the variable 'fathers_man'\n","fathers_man = \n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GEvIod9xI2v1"},"source":["# And now assign the data for P. polionotus fathers to the variable 'fathers_pol'\n","fathers_pol = \n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"vuxzSgMmJSq7"},"source":["# Assign the data for P. maniculatus mothers to the variable 'mothers_man'\n","mothers_man = \n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"0rTedXMZJTW8"},"source":["# Assign the data for P. maniculatus mothers to the variable 'mothers_pol'\n","mothers_pol = \n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lWn99YBzJIH_"},"source":["### **Execute** the following code cell to make a new dataframe from the variables you just created."]},{"cell_type":"code","metadata":{"id":"BthgH9F7Iujn"},"source":["df_by_sex = pd.DataFrame({\n","    'female' : np.asarray(list(mothers_man) + list(mothers_pol)),\n","    'male' : np.asarray(list(fathers_man) + list(fathers_pol))\n","  })"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"NUxyYqWdgTXv"},"source":["# use the seaborn boxplot( ) function to plot the data in this new dataframe 'df_by_sex' \n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Klo0b14q7WGk"},"source":["### The authors reported a significant effect of sex for all of the behaviors (mothers and fathers differed significantly in their performance of each behavior). If this was such a significant effect, why do you think the authors did not put the plot you just made in the figure? "]},{"cell_type":"markdown","metadata":{"id":"I_tdeQ-b7wBr"},"source":["__Type your answer__ in this cell markdown cell\n"]},{"cell_type":"markdown","metadata":{"id":"WLQz1EAky_ZE"},"source":["-----\n","<br>\n","\n","# <font color=green> Figure 2 </font>"]},{"cell_type":"code","metadata":{"id":"fbl33r_U0R3o","cellView":"form"},"source":["#@title __execute__ this cell to import and format data for Figure 2\n","\n","# URL of data to download\n","data_url = 'https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM112_ESM.xlsx'\n","\n","# Get the data and save it locally as \"data.xls\"\n","data, headers = urllib.request.urlretrieve(data_url, './data.xls')\n","\n","xls = pd.ExcelFile(data)\n","\n","df = pd.read_excel(xls, sheet_name = 'panel C', header = None, skiprows = 3,names=['', '', '','Not_Cross_Fostered','Cross_Fostered','',''])\n","df_maniculatus_father_licking = df.reset_index()[['Not_Cross_Fostered','Cross_Fostered']]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zczzBwoYkDJ2"},"source":["### The authors said they found that there was a significant difference in the licking of *P. maniculatus* fathers that were cross-fostered versus those that were not. ***But*** they did not report the mean licking value in each condition. Let's calculate that now.\n","\n","### By executing the last cell, you imported a dataframe called `df_maniculatus_father_licking` that contains data for the licking behavior of *P. maniculatus* fathers. There are two columns: \"*Not_Cross_Fostered*\" and \"*Cross_Fostered*\"\n","\n","> You *could* get the mean of each column using the following method: <br>\n","`np.mean(dataframe_name['column_name'])`. <br> *However*, Dataframes have a bunch of nifty methods/functions associated with them that make this kind of calculation really easy. The [`mean( )`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html) method is one of them . Note that you don't even have to give this function arguments in order to use it. Apply this function to the dataframe `df_maniculatus_father_licking` below: "]},{"cell_type":"code","metadata":{"id":"kX_fa4_Au83-"},"source":["# apply the mean() method to the df_maniculatus_father_licking dataframe\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SrPBKouOwB4i"},"source":["### Also, the data for this comparison is very visually 'smushed' in the Figure. \n","\n","### Use seaborn's boxplot() function that you just learned to plot the data in df_maniculatus_father_licking.\n","\n","\n","\n"]},{"cell_type":"code","metadata":{"id":"iUAeZc7HwNkQ"},"source":["# use seaborn to plot the data from the df_maniculatus_father_licking dataframe here\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ywx9P4fa90iO"},"source":["### To test whether there is a significant difference between the behavior of two groups, one method researchers often use is the statistical '[t-test](https://en.wikipedia.org/wiki/Student%27s_t-test)'.\n","* python has many packages for statistical testing. [Scipy.stats](https://docs.scipy.org/doc/scipy/reference/stats.html) is one of them.\n","* [ttest_ind](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html#scipy.stats.ttest_ind) performs an independent students t-test"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"l9aieTWfy_wE","executionInfo":{"status":"ok","timestamp":1613679893628,"user_tz":300,"elapsed":216,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}},"outputId":"d42a0235-fd1e-48c1-933e-3173d183c425"},"source":["# Run this cell to test the null hypothesis \n","# that the results under the two test conditions are the same.\n","scipy.stats.ttest_ind(df_maniculatus_father_licking['Not_Cross_Fostered'],\n","                      df_maniculatus_father_licking['Cross_Fostered'],\n","                      nan_policy = 'omit')"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["Ttest_indResult(statistic=2.3038228486155976, pvalue=0.027669007620192)"]},"metadata":{"tags":[]},"execution_count":63}]},{"cell_type":"markdown","metadata":{"id":"J58c-sVo0f5p"},"source":["### How would you state the results of this t-test in terms of the p-value result? <bR>(Don't forget to note the 'significance level') "]},{"cell_type":"markdown","metadata":{"id":"U7M4zIB90zfy"},"source":["***type*** your answer in this markdown"]},{"cell_type":"markdown","metadata":{"id":"bDtyO_FX7Lln"},"source":["---\n","\n","<br>\n","\n","# <font color=green> Figure 3 : Section I</font>"]},{"cell_type":"code","metadata":{"id":"eYmoGjLfJiio","cellView":"form","executionInfo":{"status":"ok","timestamp":1613759356188,"user_tz":300,"elapsed":975,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}}},"source":["#@title __execute__ this cell to import and format data for Figure 3\n","\n","# URL of data to download\n","data_url = 'https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM114_ESM.xlsx'\n","\n","# Get the data and save it locally as \"data.xls\"\n","data, headers = urllib.request.urlretrieve(data_url, './data.xls')\n","\n","xls = pd.ExcelFile(data)\n","\n","df = pd.read_excel(xls, sheet_name = 'panels B-E', header = None, skiprows = 1,names=['ID', 'Parent', 'Species','Retrieval','Huddling','Licking','NestQuality'])"],"execution_count":16,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"sQpoFGz29bOi"},"source":["### Take a moment to look at the dataframe for Figure 3. \n","### Notice how it is organized it a bit differently than the dataframes from the other figures. "]},{"cell_type":"code","metadata":{"id":"dr8EcRb974qo"},"source":["# display the contents of the dataframe for Figure 3 'df'\n","df"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tqfqASUWAWvu"},"source":["### Each row corresponds to a different individual. Each column contains information about each individual. "]},{"cell_type":"markdown","metadata":{"id":"_g-2t2w8AYEe"},"source":["### We can get a list of criteria for each column by getting the \"*unique*\" values in that column. \n","  * The [`unique( )`](https://numpy.org/doc/stable/reference/generated/numpy.unique.html) function is part of the numpy package (which was imported as \"`np`\")\n","  * Remember that dataframe columns are indexed by their column name "]},{"cell_type":"code","metadata":{"id":"ApkujvQgAyhD"},"source":["# use the np.unique() function to print the list of criteria in the 'Species' column\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KGkAsVzT97wF"},"source":["### We can find individuals fulfilling certain criteria by asking which rows are equal to (```==```) that certain criteria in a given column. For example, to find which individuals are mothers we would type ```df['Parent']=='mother'```. In other words, we are stating \"**the parent is a mother**.\" In return, we get an assessment of whether this statement was true or false for each row. \n","### Try this in the cell below (note capitalization in column headers):"]},{"cell_type":"code","metadata":{"id":"YKEsSVAD97I9"},"source":["\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"223piZgm-1wB"},"source":["### For each dataframe row in which the statement \"The parent is female\" is true, there is a ' ```True``` ' and vice versa for each row in which the statement is false. The result is called a *boolean* array and is very useful for sorting through the dataframe to isolate information we are interested in.  <br>\n","\n","### In the following cell, \n","* 1) assign the expression above to a variable called '`bool_mother`' <br>\n","* 2) use that variable (a boolean array) to index the original dataframe \n","> this isolates the data for just mothers from the original dataframe\n","* 3) assign the indexed original dataframe to a new one called **df_mothers**."]},{"cell_type":"code","metadata":{"id":"YwI5lmmzAFHD","executionInfo":{"status":"ok","timestamp":1613753630649,"user_tz":300,"elapsed":382,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}}},"source":["# Place your code here\n","\n"],"execution_count":6,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"nssPbFubAS1h"},"source":["#### Great! If you compared the size (```len()```) of *df_mothers* to *df*, you would see that *df_mothers* contains data from a smaller number of animals (because it leaves out all data for fathers).<br>\n","> ##### Use the code cell below to look at this for yourself (refer to earlier in the notebook for examples of how you examined the length of a dataframe before). <br>"]},{"cell_type":"code","metadata":{"id":"I4_2wmSEAYPp"},"source":["# compare the length of df_mothers to the length of df\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"9XUt3sPe6LV_"},"source":["# The authors use Figure 3 to make the following statement: <br>\n","> \"*The distribution of each component of parental care among the F2 mice encompassed the distributions of both species (Fig. 3bâ€“e). On the basis of the largely unimodal distributions of parental behaviours among the F2 hybrids, which resembled P. maniculatus more closely than P. polionotus, the more extensive parental care of P. polionotus probably involves more than one genetic locus.*\" <br>\n","### We can plot the data from Figure 3 in a different way that may help us visualize this better. We will visually compare the behavior data for all mothers in the F2 generation to the behavior data for mothers of each of the two species in the Parental generation individually. To do so, we will make smoothed histogram for each group and overlay them on the same plot.\n","\n","* Create a boolean array for each Parent species and assign these booleans to the variables: <br>\n"," ' ```bool_polionotus``` ' and ' ```bool_maniculatus``` '\n"," > Take care to use the *df_mothers* dataframe you had created rather than the original *df* dataframe <br>\n","* Create a boolean array for the 'F2' species and assign this boolean to the variable: <br>\n"," ' ```bool_F2``` '\n"," \n","* Pick a behavior to plot (Retrieval, Huddling, Licking, or NestQuality).\n","* Index the df_mothers dataframe you just created to extract the data column corresponding to the behavior you chose. \n","* Seaborn has a [kdeplot( )](https://seaborn.pydata.org/generated/seaborn.kdeplot.html#seaborn.kdeplot) function that you will use to plot the *distribution* of this data. This function creates a ***smoothed histogram plot***. \n","> *KDE* stands for 'kernel density estimate'"]},{"cell_type":"code","metadata":{"id":"sW2kpathMkKI","executionInfo":{"status":"ok","timestamp":1613759968410,"user_tz":300,"elapsed":386,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}}},"source":["# Create a the boolean array for P. polionotus\n","bool_polionotus = \n","\n","# Create a the boolean array for P. polionotus\n","bool_maniculatus = \n","\n","# Create a the boolean array for F2\n","bool_F2 = \n"],"execution_count":18,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"e7SQb9qJHfD5"},"source":["\n","### Use the seaborn kdeplot( ) function to plot the behavior data for :\n","1. Individuals for which bool_polionotus is True\n","2. Individuals for which bool_maniculatus is True\n","3. Individuals for which bool_F2 is True\n","> ***hint*** index your dataframe using the booleans you created <br>\n","* Make sure to include the code for all three plots in the ***same code cell*** so that the distributions will be overlaid in the same plot.\n","* Include a 'color' argument in the function and specify a different color for each group. This way you can keep track of which data belongs to which group.\n","\n","### **NOTE** To include the data for only a single behavior, the dataframe needs to also be indexed by the behavior name (column name)\n","> `dataframe[boolean index][behavior_name]`\n","\n"]},{"cell_type":"code","metadata":{"id":"yvJLUa2CHsyE"},"source":["# Create a kdeplot( ) for each of the three groups you are comparing\n","# for each group, plot data from only one behavior of interest.\n","sns.kdeplot()\n","sns.kdeplot()\n","sns.kdeplot()\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tM_m0-WXmt7U"},"source":["### Compare your plot to the plots in Figure 3 of the paper. Where is the data for each of the lines you just plotted located in that figure? In the plot you just made, you can more explicitly examine overlap between the distributions of the different groups."]},{"cell_type":"markdown","metadata":{"id":"6APFNPvdu3E6"},"source":["---\n","\n","<br>\n","\n","# <font color=green> Figure 3 : Section II</font>"]},{"cell_type":"markdown","metadata":{"id":"yy1_rmj-fs4A"},"source":["# The authors noted that some of the parental behaviors were correlated with each other.\n","### What does that mean in other words? Describe using an example from the data in this paper. "]},{"cell_type":"markdown","metadata":{"id":"e-6DQDG_fuld"},"source":["**Type** your answer in this markdown "]},{"cell_type":"markdown","metadata":{"id":"AWg8Qk3cgNh_"},"source":["### The authors tested for correlations among behaviours in the F2 progeny. Let's compare that to correlations among behaviors in the Parental animals using the data from Figure 3.\n","* Look at either mothers or fathers by indexing the Figure 3 dataframe (`df`) by a boolean for df['Species']==mothers/fathers.\n","* Assign the output to a new dataframe variable called 'df_mothers' or 'df_fathers'\n","* The [`corr( )`](https://www.geeksforgeeks.org/python-pandas-dataframe-corr/) method of a dataframe calculates the correlations among all of its columns containing numerical values. <br> Use the `corr( )` method to calculate behavior correlations for:\n"," * The parental generation of one species\n"," * The F2 generation "]},{"cell_type":"code","metadata":{"id":"QvBqh4cisdxi","executionInfo":{"status":"ok","timestamp":1613761652247,"user_tz":300,"elapsed":357,"user":{"displayName":"Krista Perks","photoUrl":"","userId":"11118389411424303278"}}},"source":["# create a dataframe containing only mothers or only fathers\n","\n","\n","# create a boolean to index all individuals in the F2 generation\n","\n","\n","# create a boolean to index all individuals \n","# in the Parent generation of one species (P. polionotus or P. maniculatus)\n","\n"],"execution_count":27,"outputs":[]},{"cell_type":"code","metadata":{"id":"e5ZzGz3gsK7W"},"source":["# calculate the correlation among behaviors in the F2 generation\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"RixJCUWvsPAu"},"source":["# calculate the correlation among behaviors in the Parent generation\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"wGlxXl6psVeg"},"source":["### How do the correlations compare between these two groups?"]},{"cell_type":"markdown","metadata":{"id":"td1-f71DMGE3"},"source":["### We can visualize correlations using a scatter plot with one variable on the x-axis and the other variable on the y-axis. You have already learned how to use the [`plt.scatter( )`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.scatter.html) function.\n","### Which behaviors are the most and least correlated in the F2 generation? \n","* Make a scatter plot of the two most correlated behaviors against each other \n","* Make a scatter plot of the least two correlated behaviors against each other."]},{"cell_type":"code","metadata":{"id":"H3QoVitZMHlv"},"source":["# make a scatter plot of the two most correlated behaviors against each other\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tdaOpeZZIFo1"},"source":["### Does it look like you expected? Why or why not?"]},{"cell_type":"code","metadata":{"id":"rlT9fMNjH0Ji"},"source":["# make a scatter plot of the two least correlated behaviors against each other\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UpNuSyX0ctGY"},"source":["### From this analysis the authors concluded that: <br>\n","> \" *Some genetic loci affect multiple parental  behaviours, whereas nest building is more genetically independent from the other behaviours measured.* \"\n","### __Discuss__ this conclusion and how the correlations you just calculated and visualized supports their statement. What would be different about the correlations you just calculated if the alternative (that all behaviors are genetically independent from each other) were true?\n","\n"]},{"cell_type":"markdown","metadata":{"id":"M_icFWfDJvGb"},"source":["***your answer here***"]},{"cell_type":"markdown","metadata":{"id":"iTIBzPxXF4yy"},"source":["---\n","\n","<br>\n","\n","# <font color=green> Figure 4 </font>"]},{"cell_type":"markdown","metadata":{"id":"LMwDsaj9nP-e"},"source":["### Figure 4 proviides a visualization of the correlation coefficient table for fathers as well as mothers. The authors note that there is a stronger effect in the fathers. \n","* Does your result match the one reported in the paper? We will have to load their source data for the figure to compare since they color-coded the results rather than providing the explicit quantifications.\n"," > (Note that in Figure 4, they also analyzed two other behaviors (Handling and Approach) that we do not have data for in Figure 3.)"]},{"cell_type":"code","metadata":{"id":"HQ4y7VrJj9uu","cellView":"form"},"source":["#@title **execute** to load and display the raw correlation coefficient values plotted in Figure4\n","\n","# URL of data to download\n","data_url = 'https://static-content.springer.com/esm/art%3A10.1038%2Fnature22074/MediaObjects/41586_2017_BFnature22074_MOESM116_ESM.xlsx'\n","\n","# Get the data and save it locally as \"data.xls\"\n","data, headers = urllib.request.urlretrieve(data_url, './data.xls')\n","\n","xls = pd.ExcelFile(data)\n","\n","df_Fig4_fathers = pd.read_excel(xls,nrows=6,skiprows=3,header=None,names=['Nest quality', 'Time licking pup', 'Time huddling pup','Fraction of pups retrieved','Promptness to handle pup','Promptness to approach pup'])\n","df_Fig4_mothers = pd.read_excel(xls,nrows=6,skiprows=12,header=None,names=['Nest quality', 'Time licking pup', 'Time huddling pup','Fraction of pups retrieved','Promptness to handle pup','Promptness to approach pup'])\n","from termcolor import colored\n","print('')\n","print(\"\\033[31m \\033[1m\" + 'Figure 4 -- F2 Mothers:' + \"\\033[0m\")\n","display(df_Fig4_mothers)\n","print('')\n","print('')\n","print('')\n","print(\"\\033[31m \\033[1m\" + 'Figure 4 -- F2 Fathers:' + \"\\033[0m\")\n","display(df_Fig4_fathers)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3knWfRma4rbG"},"source":["---\n"]},{"cell_type":"markdown","metadata":{"id":"9n29F76VQLii"},"source":["## Resources\n","For additional Jupyter Notebook information and practice, see [this tutorial](https://www.dataquest.io/blog/jupyter-notebook-tutorial/) from DataQuest. \n","\n","## About this Notebook\n","This notebook was created by [Krista Perks](https://github.com/neurologic) for the Animal Behavior class at Wesleyan University."]}]}